robot/client/client.go

```diff
@@ -1190,6 +1192,64 @@ func (rc *RobotClient) Version(ctx context.Context) (robot.VersionResponse, erro
 	return mVersion, nil
 }
 
+// Tunnel tunnels data to/from the read writer from/to the destination port on the server. This
+// function will close the connection passed in as part of cleanup.
+func (rc *RobotClient) Tunnel(ctx context.Context, conn net.Conn, dest int) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	client, err := rc.client.Tunnel(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := client.Send(&pb.TunnelRequest{
+		DestinationPort: uint32(dest),
+	}); err != nil {
+		return err
+	}
+	rc.Logger().CDebugw(ctx, "creating tunnel to server", "port", dest)
+	var (
+		wg              sync.WaitGroup
+		readerSenderErr error
+	)
+	wg.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer func() {
+			// we communicate an end to the stream by calling CloseSend(),
+			// which is why we don't need an EOF field on the request message.
+			client.CloseSend()
+			// By cancelling this ctx, we will close the client, meaning client.Recv() in the RecvWriterLoop will exit
+			// and return an error.
```

Comment:
Particularly because of the `no stream for id; discarding` error, I'm inclined to ask: are there any other synchronization mechanisms between the reader-sender and recv-writer goroutines besides a context cancelation? I buy that this works, but it's a little odd to force an error on the underlying bidi stream (cancel the context) rather than use a channel or something to communicate a required exit between the two goroutines.

Reply:
Thinking if there's an alternative to mucking with our logs: would instead scheduling a `time.AfterFunc(5*time.Second, cancel)` help. Such that we only log when our closing takes > 5 seconds to "roundtrip".

IIUC we're properly propagating closing up and down the tunnel, but we just don't give it time to politely disconnect.

Reply:
added an after func - with the current set of changes we do not get the `no stream for id` error at all, and usually everything cleans up before the 5 second mark